source/agora/consensus/EnrollmentManager.d

```diff
@@ -1093,3 +1107,103 @@ unittest
         }
     }
 }
+
+/// tests for `EnrollmentManager.getValidatorCount
+unittest
+{
+    import agora.common.Amount;
+    import agora.consensus.data.Transaction;
+    import agora.consensus.Genesis;
+    import std.algorithm;
+    import std.format;
+    import std.conv;
+
+    scope storage = new TestUTXOSet;
+
+    auto gen_key_pair = getGenesisKeyPair();
+    KeyPair key_pair = KeyPair.random();
+
+    foreach (idx; 0 .. 8)
+    {
+        auto input = Input(hashFull(GenesisTransaction), idx.to!uint);
+
+        Transaction tx =
+        {
+            TxType.Freeze,
+            [input],
+            [Output(Amount.MinFreezeAmount, key_pair.address)]
+        };
+
+        auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        storage.put(tx);
+    }
+
+    // create an EnrollmentManager object
+    auto man = new EnrollmentManager(":memory:", key_pair);
+    scope (exit) man.shutdown();
+    Hash[] utxo_hashes = storage.keys;
+
+    Enrollment enrollment;
+    ulong block_height = 2;
+
+    // create and add the first Enrollment object
+    auto utxo_hash1 = utxo_hashes[0];
+    assert(man.createEnrollment(utxo_hash1, block_height, enrollment));
+    assert(!man.hasEnrollment(utxo_hash1));
+    assert(man.addEnrollment(enrollment, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash1));
+    assert(man.getValidatorCount(block_height) + 1 == 1);
+
+    man.clearExpiredValidators(block_height);
+    man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    block_height = 3;
+
+    // create and add the second Enrollment object
+    auto utxo_hash2 = utxo_hashes[1];
+    assert(man.createEnrollment(utxo_hash2, block_height, enrollment));
+    assert(!man.hasEnrollment(utxo_hash2));
+    assert(man.addEnrollment(enrollment, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash2));
+    assert(man.getValidatorCount(block_height) + 1 == 2);
+
+    man.clearExpiredValidators(block_height);
+    man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    block_height = 4;
+
+    // create and add the third Enrollment object
+    auto utxo_hash3 = utxo_hashes[2];
+    assert(man.createEnrollment(utxo_hash3, block_height, enrollment));
+    assert(!man.hasEnrollment(utxo_hash3));
+    assert(man.addEnrollment(enrollment, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash3));
+    assert(man.getValidatorCount(block_height) + 1 == 3);
+
+    man.clearExpiredValidators(block_height);
+    man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    block_height = 5;    // valid block height : [0 .. 1008]
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 3);
+
+    block_height = 1009; // valid block height : [2 .. 1009]
```

Comment:
I can see two options here:
```suggestion
    block_height = 1009; // valid block height : [2, 1009]
```
OR:
```suggestion
    block_height = 1009; // valid block height : [2 .. 1010]
```

I have a slight preference for the second one, but can see how it could be confusing.
But if we are using the D syntax for array subscription, we have to make it non-inclusive at the end.
In mathematical notation, we would use `[)`, not `[]`, for this.